File: capture/android/PressurePulsationsRecorder/src/main/java/com/avyss/PressurePulsationsRecorder/acquisition/PressureCollectingListener.java
package com.avyss.PressurePulsationsRecorder.acquisition;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;

import java.util.function.Consumer;

public class PressureCollectingListener extends AbstractSampleCollector implements SensorEventListener {

    private final Consumer<Integer> progressListener;

    private int lastSecondsProgress;

    public PressureCollectingListener(float samplesPerSecond, int maxRecordingLengthSec, long recStartTimeNanos, Consumer<Integer> progressListener) {
        super(samplesPerSecond, maxRecordingLengthSec, recStartTimeNanos);

        this.progressListener = progressListener;
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if ( event.sensor.getType() != Sensor.TYPE_PRESSURE) {
            return;
        }

        Float elapsedSeconds = onSampleAcquired(event.values[0], event.timestamp);

        if (elapsedSeconds != null) {
            int currSecondsProgress = (int) Math.floor(elapsedSeconds);
            if (lastSecondsProgress != currSecondsProgress) {
                lastSecondsProgress = currSecondsProgress;
                progressListener.accept(currSecondsProgress);
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }
}

File: capture/android/PressurePulsationsRecorder/src/main/java/com/avyss/PressurePulsationsRecorder/acquisition/PressureCollectingListener.kt
package com.avyss.PressurePulsationsRecorder.acquisition

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import com.avyss.PressurePulsationsRecorder.data.NamedExportableData
import com.avyss.PressurePulsationsRecorder.data.NamedExportableValuesLine
import com.avyss.PressurePulsationsRecorder.data.RateAccommodatingSampleCollector

class PressureCollectingListener(
        private val samplesPerSecond: Float,
        maxRecordingLengthSec: Int,
        recStartTimeNanos: Long
): SensorEventListener {

    private val sampleCollector = RateAccommodatingSampleCollector(
            samplesPerSecond,
            maxRecordingLengthSec,
            recStartTimeNanos,
            1)

    companion object {
        val SAMPLES_COLUMNS_NAMES = arrayOf("time", "pressure")
        val SAMPLING_RATE_COLUMNS_NAMES = arrayOf("sampling_rate")
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type != Sensor.TYPE_PRESSURE) {
            return
        }

        sampleCollector.onSampleAcquired(event.timestamp, event.values[0])
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}

    fun exportableSamples(): NamedExportableData {
        return sampleCollector.getExportable().withNames(SAMPLES_COLUMNS_NAMES)
    }

    fun exportableFs(): NamedExportableData {
        return NamedExportableValuesLine(SAMPLING_RATE_COLUMNS_NAMES, floatArrayOf(samplesPerSecond))
    }

}
